packages/bitcore-node/src/modules/ripple/p2p.ts

```diff
@@ -0,0 +1,332 @@
+      logger.warn(
+        `${timestamp()} | Not connected to peer: ${host}:${port || ''} | Chain: ${this.chain} | Network: ${
+          this.network
+        }`
+      );
+    });
+    this.events.on('connected', async () => {
+      this.client!.on('ledger', async () => {
+        this.sync();
+      });
+    });
+  }
+
+  async disconnect() {
+    this.disconnecting = true;
+  }
+
+  async handleReconnects() {
+    this.disconnecting = false;
+    let firstConnect = true;
+    let connected = false;
+    let disconnected = false;
+    const { host, port } = this.chainConfig.provider;
+    while (!this.disconnecting && !this.stopping) {
+      try {
+        try {
+          this.client = await this.provider.getClient(this.network);
+          connected = this.client.isConnected();
+        } catch (e) {
+          connected = false;
+        }
+        if (connected) {
+          if (disconnected || firstConnect) {
+            this.events.emit('connected');
+          }
+        } else {
+          this.client = await this.provider.getClient(this.network);
+          this.events.emit('disconnected');
+        }
+        if (disconnected && connected && !firstConnect) {
+          logger.warn(
+            `${timestamp()} | Reconnected to peer: ${host}:${port || ''} | Chain: ${this.chain} | Network: ${
+              this.network
+            }`
+          );
+        }
+        if (connected && firstConnect) {
+          firstConnect = false;
+          logger.info(
+            `${timestamp()} | Connected to peer: ${host}:${port || ''} | Chain: ${this.chain} | Network: ${
+              this.network
+            }`
+          );
+        }
+        disconnected = !connected;
+      } catch (e) {}
+      await wait(5000);
+    }
+  }
+
+  async connect() {
+    this.handleReconnects();
+  }
+
+  public async getBlock(height: number) {
+    return this.provider.getBlock({ chain: this.chain, network: this.network, blockId: height.toString() });
+  }
+
+  async syncWallets() {
+    return new Promise(async resolve => {
+      try {
+        const { chain, network } = this;
+        const count = await WalletAddressStorage.collection.countDocuments({ chain, network });
+        let done = 0;
+        logger.info(`Syncing ${count} ${chain} ${network} wallets`);
+        let lastLog = Date.now();
+        const addressStream = WalletAddressStorage.collection.find({ chain, network }).stream();
+        addressStream
+          .pipe(
+            new Transform({
+              objectMode: true,
+              transform: async (data, _, cb) => {
+                if (Date.now() - 5000 > lastLog) {
+                  logger.info(`Syncing ${count - done} ${chain} ${network} wallets`);
+                }
+                const walletAddress = (data as any) as IWalletAddress;
+                const [lastTx] = await XrpTransactionStorage.collection
+                  .find({ chain, network, wallets: walletAddress.wallet })
```

Comment:
need to make sure this hits the 
```
{ wallets: 1, blockTimeNormalized: 1 },
      { background: true, partialFilterExpression: { 'wallets.0': { $exists: true } } }
```
index, may even want to take out chain and network

Reply:
```
{wallets: ObjectId("5e161f16d7012a1fb9d4b7da"),  'wallets.0': { $exists: true }
```

Had to change this